### PR TITLE
Statically wire up the entire IMDS, and handle link watch

### PR DESCRIPTION
**Pull request by sargun**
There were a bunch of race conditions in the IMDS due to how certain
behaviours were being handled in the Linux kernel (for example, veth
bringup was / is async). This in turn, makes sure that's all okay.

Beyond that, it does static programming of the neighbor table,
so we don't have to take any chances on neighbor discovery
"acting weird".

### CODE REVIEW

File: metadataserver/inject/inject_linux.go [hunk 1]
```diff
@@ -338,3 +364,59 @@ done:
...
+			toIMDSLink, err = containerNSHandle.LinkByName(toimdsInterfaceName)
+			if err != nil {
+				return fmt.Errorf("Could not get %s: %w", toimdsInterfaceName, err)
+			}
+
+			toContainerLink, err = intermediateNSHandle.LinkByName(tocontainerInterfaceName)
+			if err != nil {
+				return fmt.Errorf("Could not get %s: %w", tocontainerInterfaceName, err)
+			}
+
```
Comment (solarkennedy) [inline metadataserver/inject/inject_linux.go:411]: Do we have to continuously poll to get these? 
By the time we run this wait function, is it not safe to assume the Links themselves exist?

File: metadataserver/inject/inject_linux.go [hunk 2]
```diff
@@ -338,3 +364,59 @@ done:
...
+	defer t.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, fmt.Errorf("Context complete before able to get LL addrs: %w", ctx.Err())
+		case <-t.C:
+			addrs, err := handle.AddrList(link, netlink.FAMILY_V6)
+			if err != nil {
+				return nil, fmt.Errorf("Could not list V6 addrs on %q: %w", link.Attrs().Name, err)
+			}
```
Comment (solarkennedy) [inline metadataserver/inject/inject_linux.go:379]: Is this function specific to ipv6? 
If so, can you rename this function to `getLinkLocalIPv6Addr` maybe?
I just worry sometimes when I see `addr.IP` and making a mistake thinking that it is v4 when it is actually v6.

File: metadataserver/inject/inject_linux.go [hunk 3]
```diff
@@ -326,6 +328,30 @@ done:
...
+		LinkIndex:    tocontainer.Attrs().Index,
+		Family:       netlink.FAMILY_V6,
+		State:        netlink.NUD_PERMANENT,
+		Type:         unix.RTN_UNICAST,
+		IP:           metadataserviceLinkLocal.IP,
+		HardwareAddr: metadataservice.Attrs().HardwareAddr,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("Cannot insert neighbor entry %w from imds NS to container NS", err)
+	}
```
Comment (solarkennedy) [inline metadataserver/inject/inject_linux.go:353]: I look forward to when we can do this same trick for the ipv6 router or gateway or whatever.

**Review by solarkennedy — APPROVED** [review]

File: metadataserver/inject/inject_linux.go (hunk 1 above)
Comment (sargun) [reply metadataserver/inject/inject_linux.go:411]: In order to get the current _status_ of the link, we have to refetch the data. Netlink also has a "monitor" API, but the golang library doesn't suppot it as far as I can tell.

File: metadataserver/inject/inject_linux.go (hunk 2 above)
Comment (sargun) [reply metadataserver/inject/inject_linux.go:379]: Right, that makes sense.

File: metadataserver/inject/inject_linux.go (hunk 3 above)
Comment (sargun) [reply metadataserver/inject/inject_linux.go:353]: You mean on eth0? Or...?

Comment (solarkennedy) [reply metadataserver/inject/inject_linux.go:353]: Yes, something like:

```
containerNSHandle.NeighSet(&netlink.Neigh{
  LinkIndex:    eth0.Attrs().Index,	
  Family:       netlink.FAMILY_V6,
  State:        netlink.NUD_PERMANENT,	
  Type:         unix.RTN_UNICAST,	
  IP:           ourIPv6GatewayIP.IP,	
  HardwareAddr: whateverOurIPv6RouterIsStaticallyGivenToUsFromEC2.Attrs().HardwareAddr,
}
```

And then never deal with neighbor discovery again?

Comment (sargun) [reply metadataserver/inject/inject_linux.go:353]: Getting the actual MAC of the phantom router is non-trivial :(.

**Pull request merged**